src/cls/rbd/cls_rbd.cc

```diff
@@ -2807,6 +2807,33 @@ int snapshot_set_limit(cls_method_context_t hctx, bufferlist *in,
     return -EINVAL;
   }
 
+  int max_read = RBD_MAX_KEYS_READ;
+  vector<snapid_t> snap_ids;
+  string last_read = RBD_SNAP_KEY_PREFIX;
+  bool more;
+
+  do {
+      set<string> keys;
+      rc = cls_cxx_map_get_keys(hctx, last_read, max_read, &keys, &more);
+      if (rc < 0)
+          return rc;
+
+      for (set<string>::const_iterator it = keys.begin();
+           it != keys.end(); ++it) {
+          if ((*it).find(RBD_SNAP_KEY_PREFIX) != 0)
+              break;
+          snapid_t snap_id = snap_id_from_key(*it);
+          snap_ids.push_back(snap_id);
+      }
+      if (!keys.empty())
+          last_read = *(keys.rbegin());
+  } while (more);
+
+  if(new_limit < snap_ids.size()) {
+      CLS_ERR("error setting snapshot limit: %s", cpp_strerror(rc).c_str());
+      return -EINVAL;
```

Comment:
Nit: perhaps ```-ERANGE``` and then add ```op_event->ignore_error_codes = {-ERANGE};``` to https://github.com/ceph/ceph/blob/master/src/librbd/journal/Replay.cc#L774